src/assets/TokenDetectionController.ts

```diff
@@ -96,25 +106,57 @@ export class TokenDetectionController extends BaseController<
       networkType: MAINNET,
       selectedAddress: '',
       tokens: [],
+      disabled: true,
+      chainId: NetworksChainId.mainnet,
     };
     this.initialize();
     this.getTokensState = getTokensState;
     this.getTokenListState = getTokenListState;
-    this.addTokens = addTokens;
+    this.addDetectedTokens = addDetectedTokens;
     onTokensStateChange(({ tokens }) => {
       this.configure({ tokens });
     });
 
-    onPreferencesStateChange(({ selectedAddress }) => {
-      const actualSelectedAddress = this.config.selectedAddress;
-      if (selectedAddress !== actualSelectedAddress) {
-        this.configure({ selectedAddress });
+    onPreferencesStateChange(({ selectedAddress, useTokenDetection }) => {
+      const prevDisabled = this.config.disabled;
+      const isSelectedAddressChanged =
+        selectedAddress !== this.config.selectedAddress;
+      const isTokenDetectionSupported = isTokenDetectionSupportedForNetwork(
+        this.config.chainId,
+      );
+      const isDetectionEnabled = useTokenDetection && isTokenDetectionSupported;
+      this.configure({ selectedAddress, disabled: !isDetectionEnabled });
+      if (isDetectionEnabled && (prevDisabled || isSelectedAddressChanged)) {
         this.detectTokens();
       }
     });
 
-    onNetworkStateChange(({ provider }) => {
-      this.configure({ networkType: provider.type });
+    onNetworkStateChange(async (networkState) => {
+      if (this.config.chainId !== networkState.provider.chainId) {
+        const incomingChainId = networkState.provider.chainId;
+        const isTokenDetectionSupported =
+          isTokenDetectionSupportedForNetwork(incomingChainId);
+        const isDetectionEnabled =
+          isTokenDetectionSupported && !this.config.disabled;
```

Comment:
Hmm... this looks like it would result in token detection being permanently disabled after switching to an unsupported network once. Even upon returning to a supported network, `this.config.disabled` would be `true` (it was disabled on the previous network), so it would stay disabled.

Reply:
Maybe better to keep the `disabled` state separate from whether the current network is supported or not. `disabled` currently represents whether we _want_ the controller to be running. We don't want to conflate that with whether it _can_ run.

Reply:
https://github.com/MetaMask/controllers/pull/808#discussion_r887632886